marshmallow_dataclass/__init__.py

```diff
@@ -81,88 +82,192 @@ class User:
 # Recursion guard for class_schema()
 _RECURSION_GUARD = threading.local()
 
+if sys.version_info >= (3, 10):
+    # match_args, kw_only, slots
+
+    @overload
+    def dataclass(
+        _cls: Type[_U],
+        *,
+        repr: bool = True,
+        eq: bool = True,
+        order: bool = False,
+        unsafe_hash: bool = False,
+        frozen: bool = False,
+        match_args: bool = True,
+        kw_only: bool = False,
+        slots: bool = False,
+        base_schema: Optional[Type[marshmallow.Schema]] = None,
+        cls_frame: Optional[types.FrameType] = None,
+    ) -> Type[_U]:
+        ...
 
-@overload
-def dataclass(
-    _cls: Type[_U],
-    *,
-    repr: bool = True,
-    eq: bool = True,
-    order: bool = False,
-    unsafe_hash: bool = False,
-    frozen: bool = False,
-    base_schema: Optional[Type[marshmallow.Schema]] = None,
-    cls_frame: Optional[types.FrameType] = None,
-) -> Type[_U]:
-    ...
-
-
-@overload
-def dataclass(
-    *,
-    repr: bool = True,
-    eq: bool = True,
-    order: bool = False,
-    unsafe_hash: bool = False,
-    frozen: bool = False,
-    base_schema: Optional[Type[marshmallow.Schema]] = None,
-    cls_frame: Optional[types.FrameType] = None,
-) -> Callable[[Type[_U]], Type[_U]]:
-    ...
-
+    @overload
+    def dataclass(
+        *,
+        repr: bool = True,
+        eq: bool = True,
+        order: bool = False,
+        unsafe_hash: bool = False,
+        frozen: bool = False,
+        match_args: bool = True,
+        kw_only: bool = False,
+        slots: bool = False,
+        base_schema: Optional[Type[marshmallow.Schema]] = None,
+        cls_frame: Optional[types.FrameType] = None,
+    ) -> Callable[[Type[_U]], Type[_U]]:
+        ...
 
-# _cls should never be specified by keyword, so start it with an
-# underscore.  The presence of _cls is used to detect if this
-# decorator is being called with parameters or not.
-def dataclass(
-    _cls: Type[_U] = None,
-    *,
-    repr: bool = True,
-    eq: bool = True,
-    order: bool = False,
-    unsafe_hash: bool = False,
-    frozen: bool = False,
-    base_schema: Optional[Type[marshmallow.Schema]] = None,
-    cls_frame: Optional[types.FrameType] = None,
-) -> Union[Type[_U], Callable[[Type[_U]], Type[_U]]]:
-    """
-    This decorator does the same as dataclasses.dataclass, but also applies :func:`add_schema`.
-    It adds a `.Schema` attribute to the class object
+    # _cls should never be specified by keyword, so start it with an
```

Comment:
Apparently, mypy recognizes parameters whose names start with a double-underscore as positional-only.
(This seems only to be documented in [the mypy cheatsheet](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html?highlight=positional-only#functions).  But the [typeshed stub] for `dataclasses.dataclass` uses the syntax.)

Anyway...  we should probably change `_cls` to `__cls`.

[typeshed stub]: https://github.com/python/typeshed/blob/95444479acf928c0c91cb25e0c6b5a2f8ade25e7/stdlib/dataclasses.pyi#L58